python-frontend/src/main/java/org/sonar/python/types/TypeInference.java

```diff
@@ -98,10 +99,19 @@ private static void inferTypesAndMemberAccessSymbols(FunctionDef functionDef, Py
       }
     }
 
-    if (TreeUtils.hasDescendant(functionDef, tree -> tree.is(Tree.Kind.TRY_STMT))) {
+    TryStatementVisitor tryStatementVisitor = new TryStatementVisitor(functionDef);
+    functionDef.accept(tryStatementVisitor);
+    if (tryStatementVisitor.hasTryStatement) {
       // CFG doesn't model precisely try-except statements. Hence we fallback to AST based type inference
       visitor.processPropagations(trackedVars);
       functionDef.accept(new BaseTreeVisitor() {
+        @Override
+        public void visitFunctionDef(FunctionDef visited) {
+          if (functionDef == visited) {
```

Comment:
I think that to avoid this condition we can call the visitor on the body like `functionDef.body().accept(new BaseTreeVisitor() {})` at line 107